### PR TITLE
#300: allow square brackets syntax and pass aliases map to graph

### PR DESCRIPTION
**Pull request by ollevche**
Part of #300 

This PR is not meant to be merged to main directly. There is also lots of TODOs left.

Main purpose of the PR is to allow using square brackets with `each.value` / `each.key` when referencing providers:

```hcl
provider random {
  alias = "a"
}

provider random {
  alias = "b"
}

provider random {
  alias = "c"
}

module "mod" {
  for_each = toset(["a", "b", "c"])

  source = "./mod"
  providers = {
    random.moda = random[each.value]
    random.modb = random[each.value]
  }
}

resource "random_id" "all" {
  byte_length = 10

  for_each = toset(["a", "b", "c"])

  provider = random[each.value]
}
```

With such configuration, proper list of `PassedProviderConfig` with `ProviderConfigRefMapping` will be passed to graph transformers. Also, `ProviderConfigRefMapping` will be passed for configuration of `data` and `resource` blocks.

## Target Release

1.9.0

## Checklist

- [X] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [X] I have not used an AI coding assistant to create this PR.
- [X] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [ ] I have run golangci-lint on my change and receive no errors relevant to my code.
- [ ] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### CODE REVIEW

File: internal/configs/resource.go [hunk 1]
```diff
@@ -696,39 +705,171 @@ func NewProviderConfigRefMapping(p *Provider) *ProviderConfigRefMapping {
...
+				Detail:   "",
+				Subject:  m.Alias.Range().Ptr(),
+			})
+		}
+
+		m.Aliases = map[addrs.InstanceKey]string{
+			addrs.NoKey: aliasVal.AsString(),
+		}
+
+		return nil
```
Comment (cam72cam) [inline internal/configs/resource.go:743]: return diags, don't want to drop warnings

File: internal/configs/resource.go [hunk 2]
```diff
@@ -696,39 +705,171 @@ func NewProviderConfigRefMapping(p *Provider) *ProviderConfigRefMapping {
...
 	}
 
-	if ref.Alias == "" {
-		return m
+	var diags hcl.Diagnostics
+
+	// There is no for each so we just put alias under NoKey.
+	if forEach == nil {
+		// TODO/Oleksandr: use static eval for alias expr?
```
Comment (cam72cam) [inline internal/configs/resource.go:722]: We did outline this as a possibility in the RFC, though it is a nice to have.

File: internal/configs/resource.go [hunk 3]
```diff
@@ -696,39 +705,171 @@ func NewProviderConfigRefMapping(p *Provider) *ProviderConfigRefMapping {
...
-	m.Aliases = map[addrs.InstanceKey]string{
-		addrs.NoKey: ref.Alias,
+	// TODO/Oleksandr: Is static identifier valid if we evaluate for_each for alias?
+	instances, instancesDiags := eval.Evaluate(forEach, StaticIdentifier{
+		Module:    eval.call.addr,
+		Subject:   "alias",
+		DeclRange: m.Alias.Range(),
+	})
+	diags = diags.Extend(instancesDiags)
+	if instancesDiags.HasErrors() {
```
Comment (cam72cam) [inline internal/configs/resource.go:753]: This piece is a bit interesting, if we can't evaluate for_each here, we are not allowed to use each.value in the providers expression.  However, that does not necessarily mean there's a config issue.  I glossed over this scenario in my prototype.

Ex:
```hcl
module "foo" {
  for_each = data.some.stuff # Static evaluation not allowed
  providers {
    random = random["foo"]
    other = random[local.other] # Similar question to L722 above
  }
}
```

An alternative approach would to use expr dependencies to see if for_each is needed to evaluate.

Pseudo:
```
refs = lang.ReferencesInExpr(aliasExpr)
if for_each in refs {
  try eval for_each
  if fail eval, return diags
  else, alias mapping logic
  return mapping
}
return simple alias
```

Comment (cam72cam) [reply internal/configs/resource.go:753]: Though that does depend on the question of if we want to allow raw strings or locals in the expression

Comment (ollevche) [reply internal/configs/resource.go:819]: That's an excellent point, thank you!

We discussed it with @Evi1Pumpkin and decided it makes sense to have {NoKey: "alias"} mapping for cases when there is no instance references in the alias expression.

I also added `HasAlias` and `HasInstanceRefsInAlias` shortcuts to easily check if we need to iterate over all the instance keys on the graph part.

Updates are in [c661554](https://github.com/opentofu/opentofu/pull/1892/commits/c661554a4171c400383e43537e0982d810304ef1)

**Review by cam72cam — APPROVED** [review]

**Pull request merged**